### ISSUE TITLE
[Bug]: Curl/URL module does not work with LLVM backend

### ISSUE THREAD

**Issue by jabraham17** [chapel-lang/chapel#26018]
The following code fails to compile with the LLVM backend.

```chapel
import URL;
var r = URL.openUrlReader("http://www.example.com");
```

```
$CHPL_HOME/modules/packages/Curl.chpl:159: In module 'Curl':
$CHPL_HOME/modules/packages/Curl.chpl:399: error: Could not find C variable CURLPAUSE_CONT - perhaps it is a complex macro?
```

However, when using the C backend this compiles and works properly. The root cause of this is that the LLVM backend can't have extern variables to macros (as the error message suggests), and so a fix for this issue would be to just have that support. However, I could not find a relevant existing issue for that feature request, and so instead am treating this as a bug with URL/Curl.

Chplenv
```
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: arm64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: system *
CHPL_HWLOC: bundled *
CHPL_RE2: bundled *
CHPL_LLVM: system
CHPL_AUX_FILESYS: none
```

Chplenv with C backend
```
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: clang *
CHPL_TARGET_ARCH: arm64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: system *
CHPL_HWLOC: bundled *
CHPL_RE2: bundled *
CHPL_LLVM: system
CHPL_AUX_FILESYS: none
```

**Comment by mppf** [comment]
> The root cause of this is that the LLVM backend can't have extern variables to macros (as the error message suggests)

Actually it can if they are simple enough. Either way, the issue is the same here.

**Comment by dlongnecke-cray** [comment]
When we are using dyno in production, could we work around this by having a param folded extern block?

Or is the proper workaround to relax the clang name resolution/lookup rules so it can use the macro definition to define the "extern int"?

**Comment by jabraham17** [comment]
I think the proper fix is to adjust `handleMacroExpr` in `clangUtil` to be able to handle these kinds of macros. As I understand it, there are a number of challenges with this
* a macro in C can be something like `#define MyCustomBlockHeader do { `, how does that translate to Chapel? How do we detect that case?
* Whats the type of the resulting expression?

I think the last bullet is why we have the limitation in the first place (correct me if I'm wrong @mppf), where today we just have limited heuristics for this

**Comment by mppf** [comment]
Yes, and we have the workaround of writing a `static inline` C function to evaluate a macro if it's too complex (and then making that available as an `extern proc` potentially with the same name as the macro). It's just not ideal because it requires some work when it comes up.

Indeed finding the resulting type is a big challenge. Related to that, you can have macros that take arguments, and in C we have to give those types, but that's not available information in the macro. I think that macros will always be a trouble spot for our C integration. I do think it's plausible that it could get a lot better if we start generating (some) C++ and use some of the `auto` type inference (for return types) or templates (for variable argument types). Of course that's probably a pretty big effort and it'd be reasonable to prioritize other things. (Though, generating C++ might help us with some interoperability scenarios, so there is some overlap there).

**Comment by jmag722** [comment]
Besides recompiling with a C backend, is there any quick workaround for this in the meantime?

**Comment by bradcray** [comment]
@jmag722 : I imagine applying the following patch would work for a compiler using the LLVM back-end:

```chapel
 diff --git a/modules/packages/Curl.chpl b/modules/packages/Curl.chpl
index 450e80d0600..c19ba3d5fa5 100644
--- a/modules/packages/Curl.chpl
+++ b/modules/packages/Curl.chpl
@@ -352,7 +352,13 @@ module Curl {
   private extern const CURL_READFUNC_ABORT:c_size_t;
   private extern const CURL_WRITEFUNC_PAUSE:c_size_t;
   private extern const CURLPAUSE_ALL: c_int;
-  private extern const CURLPAUSE_CONT: c_int;
+
+  extern {
+    #include <curl/curl.h>
+    static int CHPL_CURLPAUSE_CONT(void) {
+      return CURLPAUSE_CONT;
+    }
+  }
 
   // extern Curl types
 
@@ -1028,7 +1034,7 @@ module Curl {
         FD_ZERO(c_ptrTo(fdexcept));
 
         // Continue anything that is paused
-        ccode = curl_easy_pause(cc.curl, CURLPAUSE_CONT);
+        ccode = curl_easy_pause(cc.curl, Curl.CHPL_CURLPAUSE_CONT());
         if ccode != CURLE_OK then
           return EINVAL; // or something...
```

**Comment by jmag722** [comment]
Thanks @bradcray , that did get rid of the `CURLPAUSE_CONT` error message and it runs!

I'm now getting a warning about `.spack-env/view/lib/libcurl.so.4: no version information available (required by ...)` but that must be due to spack building a separate curl for llvm I think

**Comment by bradcray** [comment]
Hi @jmag722 : I'm not familiar with that error, but some googling suggests that it may occur when a system has two versions of curl installed and different versions are used in different contexts.  Could that be happening here?  (I don't have any experience using curl on systems with multiple versions installed).

**Comment by jmag722** [comment]
Yeah this is on me, my Spack install of llvm built its own version of curl, which seems to be conflicting with the system version

**Comment by bradcray** [comment]
OK, give a shout if you find there's anything that should/could be done on our part to reduce conflicts like this (maybe as a new issue...
